ecosystem/indexer-grpc/indexer-grpc-file-store-backfiller/src/processor.rs

```diff
@@ -0,0 +1,265 @@
+        })
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        let (sender, receiver) = tokio::sync::mpsc::channel::<Vec<Transaction>>(1000);
+        // Get the stream out.
+        let mut transactions_buffer = BTreeMap::new();
+        let mut next_version_to_process = self.starting_version;
+        let finished_starting_versions = Arc::new(Mutex::new(BTreeSet::new()));
+        let chain_id = self.chain_id;
+        let ending_version = self.ending_version;
+
+        let mut grpc_stream = self.grpc_stream.take().expect("Stream is not initialized.");
+        let init_frame = grpc_stream
+            .next()
+            .await
+            .expect("Failed to get the first frame")?
+            .response
+            .unwrap();
+        match init_frame {
+            Response::Status(signal) => {
+                if signal.r#type() != StatusType::Init {
+                    anyhow::bail!("Unexpected status signal type");
+                }
+            },
+            _ => {
+                anyhow::bail!("Unexpected response type");
+            },
+        }
+        let mut tasks = Vec::new();
+        let receiver_ref = std::sync::Arc::new(Mutex::new(receiver));
+        let file_store_operator = self.file_store_operator.clone_box();
+        for _ in 0..20 {
+            tracing::info!("Creating a new task");
+            let mut current_file_store_operator = file_store_operator.clone_box();
+            let current_finished_starting_versions = finished_starting_versions.clone();
+            let receiver_ref = receiver_ref.clone();
+            let task = tokio::spawn(async move {
+                tracing::info!("Task started");
+                loop {
+                    let transactions = {
+                        let mut receiver = receiver_ref.lock().await;
+                        // Connection may end.
+                        let transactions = match receiver.recv().await {
+                            Some(transactions) => transactions,
+                            None => return Ok(()),
+                        };
+                        // Data quality check.
+                        ensure!(transactions.len() == 1000, "Unexpected transaction count");
+                        ensure!(
+                            transactions[0].version % 1000 == 0,
+                            "Unexpected starting version"
+                        );
+                        for (ide, t) in transactions.iter().enumerate() {
+                            ensure!(
+                                t.version == transactions[0].version + ide as u64,
+                                "Unexpected version"
+                            );
+                        }
+                        transactions
+                    };
+                    let starting_version = transactions[0].version;
+                    // If uploading failure, crash the process and let k8s restart it.
+                    current_file_store_operator
```

Comment:
Do we need to throttle this somehow? I remember we ran into upload failures because we were doing it too often, or am I misremembering it was something else.

Reply:
good question, the throttling only happens to:
1. same file(most of the cases)
2. same prefix(rare, this means hotspot at gcs side)

With existing file name, we should be good, since we randomize the prefix